### PR TITLE
Fix cross-os symbol generation

### PR DESCRIPTION
**Pull request by davidwrighton**
- Use the target os to determine the symbol type, not the sdk os
- Add tests for cross os/cross arch crossgen compilation

### CODE REVIEW

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs [hunk 1]
```diff
@@ -264,6 +265,67 @@ public void It_supports_libraries_when_using_crossgen2(string targetFramework)
...
+        [RequiresMSBuildVersionTheory("17.0.0.32901")]
+        [InlineData("net6.0", "linux-x64", "windows,linux,osx", "X64,Arm64", "nonComposite", "nonselfcontained")]
+        [InlineData("net6.0", "linux-x64", "windows,linux,osx", "X64,Arm64", "composite", "selfcontained")] // Composite in .NET 6.0 is only supported for self-contained builds
+        [InlineData("net6.0", "win-x64", "windows", "X64,Arm64", "composite", "selfcontained")] // Composite in .NET 6.0 is only supported for self-contained builds
+        [InlineData("net6.0", "osx-arm64", "windows,linux,osx", "X64,Arm64", "nonComposite", "nonselfcontained")]
+        // In .NET 6.0 building targetting Windows doesn't support emitting native symbols.
+        [InlineData("net6.0", "win-x86", "windows", "X86,X64,Arm64,Arm", "nonComposite", "nonselfcontained")]
+        public void It_supports_crossos_arch_compilation(string targetFramework, string runtimeIdentifier, string sdkSupportedOs, string sdkSupportedArch, string composite, string selfcontained)
+        {
+            var projectName = $"FrameworkDependentUsingCrossArchTest";// {targetFramework}{runtimeIdentifier.Replace("-",".")}{composite}{selfcontained}";
```
Comment (AntonLapounov) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs:277]: Do we need to fix this line?

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs [hunk 2]
```diff
@@ -264,6 +265,67 @@ public void It_supports_libraries_when_using_crossgen2(string targetFramework)
             publishCommand.Execute().Should().Pass();
         }
 
+        [RequiresMSBuildVersionTheory("17.0.0.32901")]
+        [InlineData("net6.0", "linux-x64", "windows,linux,osx", "X64,Arm64", "nonComposite", "nonselfcontained")]
+        [InlineData("net6.0", "linux-x64", "windows,linux,osx", "X64,Arm64", "composite", "selfcontained")] // Composite in .NET 6.0 is only supported for self-contained builds
+        [InlineData("net6.0", "win-x64", "windows", "X64,Arm64", "composite", "selfcontained")] // Composite in .NET 6.0 is only supported for self-contained builds
+        [InlineData("net6.0", "osx-arm64", "windows,linux,osx", "X64,Arm64", "nonComposite", "nonselfcontained")]
+        // In .NET 6.0 building targetting Windows doesn't support emitting native symbols.
```
Comment (AntonLapounov) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs:273]: This comment should be applied to `win-x64` as well, correct?

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs [hunk 3]
```diff
@@ -264,6 +265,67 @@ public void It_supports_libraries_when_using_crossgen2(string targetFramework)
...
+        [InlineData("net6.0", "win-x86", "windows", "X86,X64,Arm64,Arm", "nonComposite", "nonselfcontained")]
+        public void It_supports_crossos_arch_compilation(string targetFramework, string runtimeIdentifier, string sdkSupportedOs, string sdkSupportedArch, string composite, string selfcontained)
```
Comment (AntonLapounov) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs:275]: I would use either just `It_supports_cross_compilation` or `It_supports_cross_os_and_arch_compilation`.

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs [hunk 4]
```diff
@@ -264,6 +265,67 @@ public void It_supports_libraries_when_using_crossgen2(string targetFramework)
...
+                // Running test on processor architecture that doesn't support this cross platform build
+                return;
+            }
+
+            TestProjectPublishing_Internal(projectName, targetFramework, isSelfContained: selfcontained == "selfcontained", emitNativeSymbols: true, useCrossgen2: true, composite: composite == "composite", identifier: targetFramework, runtimeIdentifier: runtimeIdentifier);
+        }
+
+        private static bool IsTargetOsOsX(string runtimeIdentifier)
+        {
+            if (runtimeIdentifier.Contains("osx") || runtimeIdentifier.Contains("macOs"))
```
Comment (AntonLapounov) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs:312]: I think `macOs` cannot be a part of a valid runtime identifier.

**Review by AntonLapounov — APPROVED** [review]
Looks good in general. Thank you!

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs [hunk 5]
```diff
@@ -307,18 +371,34 @@ public void It_supports_libraries_when_using_crossgen2(string targetFramework)
...
+                Log.WriteLine("Checking for symbol files");
+                IEnumerable<string> pdbFiles;
 
-                publishDirectory.Should().HaveFiles(new[] {
-                    GetPDBFileName(mainProjectDll, framework),
-                    GetPDBFileName(classLibDll, framework),
-                });
+                if (composite)
+                {
+                    pdbFiles = new[] { GetPDBFileName(Path.ChangeExtension(mainProjectDll, "r2r.dll"), framework, testProject.RuntimeIdentifier) };
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs:382]: The tests which set composite used to not set selfcontained. This resulted in the compiler not actually producing a composite build, so the test continued to pass. The tests have been adjusted to explicitly pass composite false now.

**Pull request merged**